src/query/functions/utils/heap.go

```diff
@@ -68,21 +70,26 @@ func NewFloatHeap(isMaxHeap bool, capacity int) FloatHeap {
 		capacity = 0
 	}
 
+	// we limit heap capacity because we don't want to go OOM
+	if capacity > limitMaxValues {
+		capacity = limitMaxValues
+	}
```

Comment:
This should be `max(capacity, seriesCount)`.

I would recommend having the signature for `takeFunc` change to `type takeFunc func(heap FloatHeap, values []float64, buckets [][]int) []float64`, and add the `k` value as a field on `takeOp`; Then in ProcessBlock, only init the heap if `k < seriesCount`, or just return the input block otherwise

Reply:
> This should be `max(capacity, seriesCount)`; if their series are large enough to be in memory, an additional heap of the same order of magnitude is negligible
> 
I guess you meant `min(capacity, seriesCount)`? 
Because `capacity` is `k` from our topk/bottomk functions, so I guess there is no point in storing more entries in the heap than we have series?

Reply:
Correct about min, and yeah; if you're asking for `topk(100, foo)` and `foo` only has 1 series, no point building a heap at all